src/rebar_config.erl

```diff
@@ -165,44 +167,50 @@ extract_pkg_hashes(Attrs) ->
                 [First|_] -> First;
                 [] -> []
             end,
-    proplists:get_value(pkg_hash, Props, []).
+     { proplists:get_value(pkg_hash, Props, []), proplists:get_value(pkg_hash_ext, Props, [])}.
 
 %% @private extract attributes from the lock file and integrate them
 %% into the full-blow internal lock format
 %% @end
 %% TODO: refine typings for lock()
--spec expand_locks(list(), list()) -> list().
-expand_locks([], _Hashes) ->
+-spec expand_locks(list(), list(), list()) -> list().
+expand_locks([], _OldHashes, _NewHashes) ->
     [];
-expand_locks([{Name, {pkg,PkgName,Vsn}, Lvl} | Locks], Hashes) ->
-    Hash = proplists:get_value(Name, Hashes),
-    [{Name, {pkg,PkgName,Vsn,Hash}, Lvl} | expand_locks(Locks, Hashes)];
-expand_locks([Lock|Locks], Hashes) ->
-    [Lock | expand_locks(Locks, Hashes)].
+expand_locks([{Name, {pkg,PkgName,Vsn}, Lvl} | Locks], OldHashes, NewHashes) ->
+    OldHash = proplists:get_value(Name, OldHashes),
+    NewHash = proplists:get_value(Name, NewHashes),
+    [{Name, {pkg,PkgName,Vsn,OldHash, NewHash}, Lvl} | expand_locks(Locks, OldHashes, NewHashes)];
+expand_locks([Lock|Locks], OldHashes, NewHashes) ->
+    [Lock | expand_locks(Locks, OldHashes, NewHashes)].
 
 %% @private split up extra attributes for locks out of the internal lock
 %% structure for backwards compatibility reasons
 -spec write_attrs(list()) -> {list(), list()}.
 write_attrs(Locks) ->
     %% No attribute known that needs to be taken out of the structure,
     %% just return terms as is.
-    {NewLocks, Hashes} = split_locks(Locks, [], []),
-    case Hashes of
+    {NewLocks, OldHashes, NewHashes} = split_locks(Locks, [], [], []),
+    case OldHashes ++ NewHashes of
```

Comment:
probably cheaper to check on:
```
case {OldHashes, NewHashes} of
    {[], []} -> {NewLocks, []};
    _ -> ...
end
```
since it won't need to append lists together only to never use the result.

Reply:
Whole heartily agree and one of my favorite patterns 👍